### ISSUE TITLE
Failing to poll data (Invalid data from coordinator)

### ISSUE THREAD

**Issue by SupaFlav** [Patrick762/hassio-bluetti-bt#10]
Hello,

First off, thank you very much for all your efforts getting this integration together!

Getting setup with my AC200M and able to poll data occasionally (between 30min to 4 hours), but unfortunately receiving errors most of the time.

```
2023-08-20 20:08:32.692 WARNING (MainThread) [custom_components.bluetti_bt.coordinator] Bleak error: failed to discover services, device disconnected
2023-08-20 20:08:32.695 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] Finished fetching Bluetti polling coordinator data in 4.228 seconds (success: True)
2023-08-20 20:08:32.695 DEBUG (MainThread) [custom_components.bluetti_bt.binary_sensor] Updating state of ac200max_ac_output
2023-08-20 20:08:32.695 ERROR (MainThread) [custom_components.bluetti_bt.binary_sensor] Invalid data from coordinator (sensor.ac200max_ac_output)
2023-08-20 20:08:32.696 DEBUG (MainThread) [custom_components.bluetti_bt.binary_sensor] Updating state of ac200max_dc_output
2023-08-20 20:08:32.696 ERROR (MainThread) [custom_components.bluetti_bt.binary_sensor] Invalid data from coordinator (sensor.ac200max_dc_output)
2023-08-20 20:08:32.698 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_dc_input_power
2023-08-20 20:08:32.699 ERROR (MainThread) [custom_components.bluetti_bt.sensor] Invalid data from coordinator (sensor.ac200max_dc_input_power)
2023-08-20 20:08:32.702 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_ac_input_power
2023-08-20 20:08:32.702 ERROR (MainThread) [custom_components.bluetti_bt.sensor] Invalid data from coordinator (sensor.ac200max_ac_input_power)
2023-08-20 20:08:32.704 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_ac_output_power
2023-08-20 20:08:32.705 ERROR (MainThread) [custom_components.bluetti_bt.sensor] Invalid data from coordinator (sensor.ac200max_ac_output_power)
2023-08-20 20:08:32.707 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_dc_output_power
2023-08-20 20:08:32.707 ERROR (MainThread) [custom_components.bluetti_bt.sensor] Invalid data from coordinator (sensor.ac200max_dc_output_power)
2023-08-20 20:08:32.708 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_total_power_generation
2023-08-20 20:08:32.708 ERROR (MainThread) [custom_components.bluetti_bt.sensor] Invalid data from coordinator (sensor.ac200max_total_power_generation)
2023-08-20 20:08:32.709 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_total_battery_percent
2023-08-20 20:08:32.709 ERROR (MainThread) [custom_components.bluetti_bt.sensor] Invalid data from coordinator (sensor.ac200max_total_battery_percent)
2023-08-20 20:08:32.710 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_ac_output_mode
2023-08-20 20:08:32.711 ERROR (MainThread) [custom_components.bluetti_bt.sensor] Invalid data from coordinator (sensor.ac200max_ac_output_mode)
2023-08-20 20:08:32.712 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_internal_ac_voltage
2023-08-20 20:08:32.712 ERROR (MainThread) [custom_components.bluetti_bt.sensor] Invalid data from coordinator (sensor.ac200max_internal_ac_voltage)
2023-08-20 20:08:32.712 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_internal_current_sensor_1
2023-08-20 20:08:32.712 ERROR (MainThread) [custom_components.bluetti_bt.sensor] Invalid data from coordinator (sensor.ac200max_internal_current_sensor_1)
2023-08-20 20:08:32.714 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_internal_power_sensor_1
2023-08-20 20:08:32.714 ERROR (MainThread) [custom_components.bluetti_bt.sensor] Invalid data from coordinator (sensor.ac200max_internal_power_sensor_1)
2023-08-20 20:08:32.715 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_internal_ac_frequency
2023-08-20 20:08:32.715 ERROR (MainThread) [custom_components.bluetti_bt.sensor] Invalid data from coordinator (sensor.ac200max_internal_ac_frequency)
2023-08-20 20:08:32.716 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_total_battery_voltage
2023-08-20 20:08:32.716 ERROR (MainThread) [custom_components.bluetti_bt.sensor] Invalid data from coordinator (sensor.ac200max_total_battery_voltage)
2023-08-20 20:08:32.717 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_screen_auto_sleep_mode
2023-08-20 20:08:32.717 ERROR (MainThread) [custom_components.bluetti_bt.sensor] Invalid data from coordinator (sensor.ac200max_screen_auto_sleep_mode)
```

If I quickly connect/disconnect from my Android device via the Bluetti app it seems to kickstart a polling interval for HA, but then reverts back to the errors above.  For infrequent successful polls from HA that can occur, I'm uncertain if my Android device is having any influence in the background.

```
2023-08-20 20:44:49.467 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] Polling data
2023-08-20 20:44:52.844 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] Requesting ReadHoldingRegisters(starting_address=10, quantity=40)
2023-08-20 20:44:53.069 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] Got 85 bytes
2023-08-20 20:44:53.069 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] Parsed data: {'device_type': 'AC200M', 'serial_number': 0, 'arm_version': Decimal('4030.15'), 'dsp_version': Decimal('4005.07'), 'dc_input_power': 0, 'ac_input_power': 470, 'ac_output_power': 0, 'dc_output_power': 0, 'power_generation': Decimal('0'), 'total_battery_percent': 22, 'ac_output_on': False, 'dc_output_on': False}
2023-08-20 20:44:53.069 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] Requesting ReadHoldingRegisters(starting_address=70, quantity=21)
2023-08-20 20:44:53.294 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] Got 47 bytes
2023-08-20 20:44:53.295 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] Parsed data: {'ac_output_mode': <OutputMode.BYPASS_OUTPUT_C: 2>, 'internal_ac_voltage': 0, 'internal_current_one': Decimal('0'), 'internal_power_one': 0, 'internal_ac_frequency': Decimal('0'), 'internal_dc_input_voltage': 0, 'internal_dc_input_power': Decimal('0'), 'internal_dc_input_current': Decimal('0')}
2023-08-20 20:44:53.295 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] Requesting ReadHoldingRegisters(starting_address=3001, quantity=61)
2023-08-20 20:44:53.510 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] Got 127 bytes
2023-08-20 20:44:53.511 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] Parsed data: {'pack_num': 1, 'ac_output_on': False, 'dc_output_on': False, 'power_off': False, 'auto_sleep_mode': <AutoSleepMode.THIRTY_SECONDS: 2>}
2023-08-20 20:44:55.447 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] Finished fetching Bluetti polling coordinator data in 5.980 seconds (success: True)
2023-08-20 20:44:55.447 DEBUG (MainThread) [custom_components.bluetti_bt.binary_sensor] Updating state of ac200max_ac_output
2023-08-20 20:44:55.447 DEBUG (MainThread) [custom_components.bluetti_bt.binary_sensor] Updating state of ac200max_dc_output
2023-08-20 20:44:55.448 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_dc_input_power
2023-08-20 20:44:55.448 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_ac_input_power
2023-08-20 20:44:55.449 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_ac_output_power
2023-08-20 20:44:55.450 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_dc_output_power
2023-08-20 20:44:55.450 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_total_power_generation
2023-08-20 20:44:55.450 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_total_battery_percent
2023-08-20 20:44:55.451 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_ac_output_mode
2023-08-20 20:44:55.451 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_internal_ac_voltage
2023-08-20 20:44:55.452 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_internal_current_sensor_1
2023-08-20 20:44:55.452 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_internal_power_sensor_1
2023-08-20 20:44:55.453 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_internal_ac_frequency
2023-08-20 20:44:55.453 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_total_battery_voltage
2023-08-20 20:44:55.453 DEBUG (MainThread) [custom_components.bluetti_bt.sensor] Updating state of ac200max_screen_auto_sleep_mode
```

The goal is obviously to have HA gather the data reliably by itself without an external device kickstarting the process.  So close to getting it working, but alas something is amiss.  Your help is greatly appreciated!

**Comment by Patrick762** [comment]
I'll look into it.
I have removed the logs because they contained you device Serial number. One other thing that needs to be changed

**Comment by SupaFlav** [comment]
Failing to understand why BleakClient connect fails most of the time, but successful once in a while at random.  Added in some logging to see what's happening, but haven't been able to sort anything out.  Originally bt receiver was ~5ft and moved within 1ft, but still having the same behavior.  For now, I've reduced the poling delay in hopes of getting a successful poll sooner (every 5s).
```
2023-08-22 12:13:29.252 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] Polling data
2023-08-22 12:13:29.252 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] XXX - Device (XX:XX:XX:XX:XX:XX: AC200MXXXXXXX) is available with type(AC200M)
2023-08-22 12:13:29.252 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] XXX - trying: client.connect()
2023-08-22 12:13:33.979 WARNING (MainThread) [custom_components.bluetti_bt.coordinator] Bleak error: failed to discover services, device disconnected
2023-08-22 12:13:33.980 DEBUG (MainThread) [custom_components.bluetti_bt.coordinator] XXX - client.disconnect()
```
If you have any thoughts or suggestions I'll give them a go.  Thanks!

**Comment by SupaFlav** [comment]
For now, it is a lot more reliable forcing the coordinator.py update_interval to 5s from 20s.  I can usually get a report between 10-30mins which is good enough for now.  I set the offending error log statements to debug to keep the home-assistant.log clean.  Until a proper fix can be implemented this will give me some confidence in catching a low or high battery level condition.

For anyone that comes across this with a similar connection issue, edit _/config/custom_components/bluetti_bt/coordinator.py_
 and set the _update_interval_ seconds to 5.  After that, you simply need to restart HA to reload the integration.

```
class PollingCoordinator(DataUpdateCoordinator):
    """Polling coordinator."""

    def __init__(self, hass: HomeAssistant, address, device_name: str):
        """Initialize coordinator."""
        super().__init__(
            hass,
            _LOGGER,
            name="Bluetti polling coordinator",
            update_interval=timedelta(seconds=5),
        )
```

**Comment by SupaFlav** [comment]
Boo ...  update intervals slowly got worse throughout the night before finally giving up.  It's been over 9 hours now and I haven't received a successful update (it's still trying though).  I'm currently running with a WAVLINK WL-BT4001 and thinking it may be the adapter.  Have a StarTech USBBT1EDR4 on order to try out tomorrow.  Fingers crossed!

**Comment by Patrick762** [comment]
You can also try the integration with a Bluetooth Proxy

**Comment by SupaFlav** [comment]
I'm hoping to just interface directly through the bluetooth adapter without adding an ESP32.  If I go the ESP32 route then I'll have other options, but wanting to limit the number of things I need to toy with.

I discovered that when I disable/enable the bluetooth connection directly at the AC200M it will allow HA to again poll successfully.  It seems the integration doesn't disconnect properly, or maybe I can just maintain the connection indefinitely.  It could also be a faulty adapter (I'll find out tomorrow).  Maybe I expected too much from a $5 adapter.

**Comment by SupaFlav** [comment]
I kludged the coordinator to maintain the connection once established and it's working really well ... so far.  Polling every 10s with only a few timeouts here and there, but the code seems to recover and polls rarely fail now.  Only upon an error do I have it disconnect and then establish a new connection.   After that, it's back to polling successfully every 10s.

Things seems to get hung up after forcing a disconnects on every poll when no fault is present, or maybe there are too many connection attempts within some period?  Haven't tried to extend the poll period to 60s or greater, or perhaps the bluetti is waiting for something and won't allow subsequent reconnects from the same client?  Don't have enough familiarity to do anything but guess at this point.

Anywho, we'll see how this goes overnight.  Would never have gotten this far without all the groundwork you've laid.  Many thanks for your work on this! 😄

**Comment by Patrick762** [comment]
I have some spare time tomorrow, so I can work on this project.
I'll try to add a poll interval option and the persistent communication option you mentioned.
Polling timeouts should also be handled as debug info and not errors, as they occur sometimes

**Comment by SupaFlav** [comment]
I created a fork and messing around. I'm a noob with regards to git, but created my own branch for the kludged persistent-connection posted above.  More of a test as it will likely get superseded by your next changes shortly.  Perhaps I can help contribute for the reporting of extra battery packs once I get a better hold of the code.

**Comment by Patrick762** [comment]
Polling interval and persistent connections are now available in release [0.0.10](https://github.com/Patrick762/hassio-bluetti-bt/releases/tag/0.0.10)
